types/wlr_pointer_gestures_v1.c

```diff
@@ -0,0 +1,274 @@
+	.destroy = resource_handle_destroy,
+};
+
+static void get_pinch_gesture(struct wl_client *client,
+		struct wl_resource *gestures_resource,
+		uint32_t id,
+		struct wl_resource *pointer_resource) {
+	struct wlr_pointer_gestures_v1 *gestures =
+		pointer_gestures_from_resource(gestures_resource);
+
+	struct wl_resource *pinch = wl_resource_create(client,
+		&zwp_pointer_gesture_pinch_v1_interface,
+		wl_resource_get_version(gestures_resource),
+		id);
+	if (pinch == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(pinch, &pinch_impl, NULL, resource_remove_from_list);
+	wl_list_insert(&gestures->pinches, wl_resource_get_link(pinch));
+}
+
+static const struct zwp_pointer_gestures_v1_interface gestures_impl = {
+	.get_swipe_gesture = get_swipe_gesture,
+	.get_pinch_gesture = get_pinch_gesture,
+};
+
+static void client_handle_destroy(struct wl_resource *resource) {
+	struct wl_client *client = wl_resource_get_client(resource);
+	struct wlr_pointer_gestures_v1 *gestures =
+		pointer_gestures_from_resource(resource);
+	struct wl_resource *gesture, *tmp = NULL;
+	wl_resource_for_each_safe(gesture, tmp, &gestures->swipes) {
+		if (wl_resource_get_client(gesture) == client) {
+			wl_resource_destroy(gesture);
+		}
+	}
+	wl_resource_for_each_safe(gesture, tmp, &gestures->pinches) {
+		if (wl_resource_get_client(gesture) == client) {
+			wl_resource_destroy(gesture);
+		}
+	}
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+static void pointer_gestures_v1_bind(struct wl_client *wl_client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_pointer_gestures_v1 *gestures = data;
+	assert(wl_client && gestures);
+
+	struct wl_resource *resource = wl_resource_create(wl_client,
+			&zwp_pointer_gestures_v1_interface, version, id);
+	if (resource == NULL) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	wl_resource_set_implementation(resource,
+			&gestures_impl, gestures, client_handle_destroy);
+	wl_list_insert(&gestures->resources, wl_resource_get_link(resource));
+}
+
+static void handle_display_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_pointer_gestures_v1 *tablet =
+		wl_container_of(listener, tablet, display_destroy);
+	wlr_pointer_gestures_v1_destroy(tablet);
+}
+
+void wlr_pointer_gestures_v1_destroy(struct wlr_pointer_gestures_v1 *gestures) {
+	struct wl_resource *resource, *tmp;
+	wl_resource_for_each_safe(resource, tmp, &gestures->resources) {
+		wl_resource_destroy(resource);
+	}
+	wl_global_destroy(gestures->global);
+	free(gestures);
+}
+
+struct wlr_pointer_gestures_v1 *wlr_pointer_gestures_v1_create(
+		struct wl_display *display) {
+	struct wlr_pointer_gestures_v1 *gestures =
+		calloc(1, sizeof(struct wlr_pointer_gestures_v1));
+	if (!gestures) {
+		return NULL;
+	}
+
+	wl_list_init(&gestures->resources);
+	wl_list_init(&gestures->swipes);
+	wl_list_init(&gestures->pinches);
+
+	gestures->display_destroy.notify = handle_display_destroy;
```

Comment:
This should be done after creating the global. If the allocation fails, there will be a dangling listener otherwise.